### PR TITLE
Snakemake

### PR DESCRIPTION
**Pull request by jmtsuji**
- Adds phylogenetic tree generation to the snakefile
- Adds `BackBLAST.sh`, a script to run the BackBLAST pipeline via snakemake from the command line. Can also generate setup templates for the user

### CODE REVIEW

File: BackBLAST.sh [hunk 1]
```diff
@@ -0,0 +1,459 @@
+#!/usr/bin/env bash
+set -euo pipefail
+# BackBLAST.sh
+# The entry script for running BackBLAST via command line
+# Copyright Lee H. Bergstrand and Jackson M. Tsuji, 2019
+
+
+# HARD-CODED variables
+VERSION=1.9.9-alpha
```
Comment (jmtsuji) [inline BackBLAST.sh]: Not sure if the version ought to be in this script.

File: BackBLAST.sh [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+SCRIPT_NAME=${0##*/}
+SCRIPT_NAME=${SCRIPT_NAME%.*}
+TEMPLATE_CONFIG="template_config.yaml"
+SNAKEFILE=Snakefile
```
Comment (jmtsuji) [inline BackBLAST.sh]: Longer-term, we'll need to figure out how to specify where these key files are in relation to the shell script (or just have them install to the same folder as the script?)

**Comment by LeeBergstrand** [comment]
@jmtsuji What's your thoughts on the shell script wrapper versus an equivalent wrapper written in Python? I think you could shorten the code a bit by writing it in Python... I think you can call snakemake directly from Python. I think that's what Atlas does. I think your shell script is beautiful though :)

File: BackBLAST.sh (hunk 1 above)
Comment (jmtsuji) [reply BackBLAST.sh]: Sounds good.

File: BackBLAST.sh [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+
+
+#' Adds the user-provided subject files to the end of the template config file
```
[the remark this answers is not in the document]
Comment (jmtsuji) [reply BackBLAST.sh]: @LeeBergstrand I like the Google style guide -- thanks for sending. Revising now.

File: BackBLAST.sh [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+    # Append to the bottom of the file (a bit hacky)
+    # Note that the arbitrary sample name is defined here for each sample as the basename of the file
+    for subject_faa_file in ${subject_faa_files[@]}; do
+
+        subject_faa_basename=${subject_faa_file%.faa}
+        subject_faa_basename=${subject_faa_basename##*/}
+        echo "  ${subject_faa_basename}: ${subject_faa_file}" >> ${output_config_filepath}
+
+    done
+
```
[the remark this answers is not in the document]
Comment (jmtsuji) [reply BackBLAST.sh]: Done

File: BackBLAST.sh [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+
+#' Runs snakemake
+#' 
+#' @param config_file: path to the config.yaml file containing settings for the run
+#' @param run_directory: path to the directory where output files should be written
+#' @param jobs: maximum number of parallel jobs for the snakemake scheduler to run
+#' @return output files from the BackBLAST pipeline, in the run_directory
+function run_snakemake {
+
```
[the remark this answers is not in the document]
Comment (jmtsuji) [reply BackBLAST.sh]: Done

File: BackBLAST.sh [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+    # Get input variables
+    local config_file=$1
+    local run_directory=$2
+    local jobs=$3
+
+    snakemake --snakefile ${SNAKEFILE} --configfile ${config_file} --directory ${run_directory} --jobs ${jobs} --use-conda --reason
+
```
[the remark this answers is not in the document]
Comment (jmtsuji) [reply BackBLAST.sh]: Done

File: BackBLAST.sh [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+        printf "   -b bootstrap_cutoff: numerical value (e.g., 80) under which bootstrap values will not be displayed [default: NA]\n"
+        printf "   -r root_name: Exact name of the tree tip label at the desired root of the tree [default: NA; will skip rooting]\n"
+        printf "   -e evalue: e-value cutoff for reciprocal BLASTP [default: 1e-40]\n"
+        printf "   -p pident: percent identity cutoff for reciprocal BLASTP [default: 25]\n"
+        printf "   -@ threads: maximum threads to use for any process [default: 1]\n\n"
+
+        # Exit
+        exit 0
+
+    fi
```
[the remark this answers is not in the document]
Comment (jmtsuji) [reply BackBLAST.sh]: Done

**Comment by jmtsuji** [comment]
> @jmtsuji What's your thoughts on the shell script wrapper versus an equivalent wrapper written in Python? I think you could shorten the code a bit by writing it in Python... I think you can call snakemake directly from Python. I think that's what Atlas does. I think your shell script is beautiful though :)

Indeed, a Python version would probably be more succinct. In retrospect, I probably should have written this in Python (I was originally thinking that the frequent file handling in the `setup` module would make Bash a better choice, but it was not as complicated as I had thought). Now that it's written, though, I'm not keen to re-write it in Python... I have a pretty big backlog of tasks right now. Would you be okay leaving it in Bash?

File: BackBLAST.sh (hunk 2 above)
Comment (jmtsuji) [reply BackBLAST.sh]: Good idea -- done

**Comment by LeeBergstrand** [comment]
> > @jmtsuji What's your thoughts on the shell script wrapper versus an equivalent wrapper written in Python? I think you could shorten the code a bit by writing it in Python... I think you can call snakemake directly from Python. I think that's what Atlas does. I think your shell script is beautiful though :)
> 
> Indeed, a Python version would probably be more succinct. In retrospect, I probably should have written this in Python (I was originally thinking that the frequent file handling in the `setup` module would make Bash a better choice, but it was not as complicated as I had thought). Now that it's written, though, I'm not keen to re-write it in Python... I have a pretty big backlog of tasks right now. Would you be okay leaving it in Bash?

@jmtsuji Ya leave it as Bash for now. I'll write it in Python if we need to.

**Comment by jmtsuji** [comment]
@LeeBergstrand Thanks for the feedback. Will merge now.

**Pull request merged**